dali/kernels/shape.h

```diff
@@ -0,0 +1,299 @@
+  //   data.resize(other.size());
+  //   for (int i = 0; i < size(); i++) {
+  //     data[i] = other[i];
+  //   }
+  //   return *this;
+  // }
+
+  // TensorShape &operator=(TensorShape<-1> &&other) {
+  //   data = std::move(other.data);
+  //   return *this;
+  // }
+
+  // We allow only explicit conversion from dynamic to static dim
+  // template <int other_dim>
+  // explicit operator TensorShape<other_dim>() const;
+  template <int other_dim>
+  TensorShape<other_dim> to_static() const;
+
+  int64_t &operator[](int d) { return shape[d]; }
+  const int64_t &operator[](int d) const { return shape[d]; }
+
+  std::vector<int64_t> shape;
+  int size() const { return shape.size(); }
+};
+
+template <int dim_>
+struct TensorShape {
+  TensorShape(const std::array<int64_t, dim_> &s) : shape(s) {
+    std::cout << "const std::array<int64_t, dim_> &s" << std::endl;
+  }
+  TensorShape(std::array<int64_t, dim_> &&s) : shape(std::move(s)) {
+    std::cout << "std::array<int64_t, dim_> &&s" << std::endl;
+  }
+
+  // We just want zero-initialization
+  TensorShape() : shape{} {}
+  // We allow only explicit operations on TensorShape static dim
+  TensorShape(const TensorShape &) = default;
+  TensorShape(TensorShape &&) = default;
+  TensorShape &operator=(const TensorShape &other) = default;
+  TensorShape &operator=(TensorShape &&other) = default;
+
+  template <typename... Ts>
+  TensorShape(int64_t i0, Ts... s) : shape{i0, int64_t{s}...} {
+    static_assert(sizeof...(Ts) == dim_ - 1, "Number of shapes passed must match dim_");
+    std::cout << "TensorShape(int64_t i0, Ts... s)" << std::endl;
+    // TODO static_assert for Ts == int64_t
+  }
+
+  int64_t &operator[](int d) { return shape[d]; }
+  const int64_t &operator[](int d) const { return shape[d]; }
+
+  std::array<int64_t, dim_> shape;
+  constexpr int size() const { return dim_; }
+};
+
+// template<int other_dim>
+// TensorShape<DynamicTensorShape>::operator TensorShape<other_dim>() const {
+//   std::cout << "operator TensorShape<other_dim>()" << std::endl;
+//   assert(size() == other_dim);
+//   TensorShape<other_dim> shape;
+//   for (int i = 0; i < other_dim; i++) {
+//     shape[i] = (*this)[i];
+//   }
+//   return shape;
+// }
+
+template <int other_dim>
+TensorShape<other_dim> TensorShape<DynamicTensorShape>::to_static() const {
+  std::cout << "operator TensorShape<other_dim>()" << std::endl;
+  assert(size() == other_dim);
+  TensorShape<other_dim> shape;
+  for (int i = 0; i < other_dim; i++) {
+    shape[i] = (*this)[i];
+  }
+  return shape;
+}
+
+// TODO should we be able to only output a tensor of sample_dim dim for all cases of TensorListShape
+template <int sample_dim_>
+struct TensorListShape {
+  TensorListShape(std::vector<TensorShape<sample_dim_>> sample_shapes)
+      : shapes(std::accumulate(sample_shapes.begin(), sample_shapes.end(), std::vector<int64_t>(),
+                               [](std::vector<int64_t> current, const TensorShape<sample_dim_> &next) {
```

Comment:
Off-topic: It's a nice example demonstrating why STL needs ranges ;) This piece of code is hardly readable, the meaning diluted with all those repetitions and calls to begin/end.
On-topic: as I pointed out above, it isn't particularly efficient, and we want efficiency dearly here. While the shape isn't large (at least in DL standards), reallocating it many times, as will most likely happen here, isn't particularly well suited for an object deemed lightweight and likely to be constructed many times.

Reply:
Off-topic: It was just to put something here, I know that is not a great peace of code and intend to make a function with clear implementation for it.

Reply:
Done.